core/src/test/scala/org/apache/spark/scheduler/ExecutorBlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,158 @@
+
+import scala.collection.mutable
+
+import org.scalatest.{BeforeAndAfter, PrivateMethodTester}
+
+import org.apache.spark._
+import org.apache.spark.scheduler.ExecutorBlacklistTracker.ExecutorFailureStatus
+import org.apache.spark.scheduler.cluster.ExecutorInfo
+import org.apache.spark.util.ManualClock
+
+class ExecutorBlacklistTrackerSuite
+  extends SparkFunSuite
+  with LocalSparkContext
+  with BeforeAndAfter {
+  import ExecutorBlacklistTrackerSuite._
+
+  before {
+    if (sc == null) {
+      sc = createSparkContext
+    }
+  }
+
+  after {
+    if (sc !=  null) {
+      sc.stop()
+      sc = null
+    }
+  }
+
+  test("add executor to blacklist") {
+    // Add 5 executors
+    addExecutors(5)
+    val tracker = sc.executorBlacklistTracker.get
+    assert(numExecutorsRegistered(tracker) === 5)
+
+    // Post 5 TaskEnd event to executor-1 to add executor-1 into blacklist
+    (0 until 5).foreach(_ => postTaskEndEvent(TaskResultLost, "executor-1"))
+
+    assert(tracker.getExecutorBlacklist === Set("executor-1"))
+    assert(executorIdToTaskFailures(tracker)("executor-1").numFailures === 5)
+    assert(executorIdToTaskFailures(tracker)("executor-1").isBlackListed === true)
+
+    // Post 10 TaskEnd event to executor-2 to add executor-2 into blacklist
+    (0 until 10).foreach(_ => postTaskEndEvent(TaskResultLost, "executor-2"))
+    assert(tracker.getExecutorBlacklist === Set("executor-1", "executor-2"))
+    assert(executorIdToTaskFailures(tracker)("executor-2").numFailures === 10)
+    assert(executorIdToTaskFailures(tracker)("executor-2").isBlackListed === true)
+
+    // Post 5 TaskEnd event to executor-3 to verify whether executor-3 is blacklisted
+    (0 until 5).foreach(_ => postTaskEndEvent(TaskResultLost, "executor-3"))
+    // Since the failure number of executor-3 is less than the average blacklist threshold,
+    // though exceed the fault threshold, still should not be added into blacklist
+    assert(tracker.getExecutorBlacklist === Set("executor-1", "executor-2"))
+    assert(executorIdToTaskFailures(tracker)("executor-3").numFailures === 5)
+    assert(executorIdToTaskFailures(tracker)("executor-3").isBlackListed === false)
+
+    // Keep post TaskEnd event to executor-3 to add executor-3 into blacklist
+    (0 until 2).foreach(_ => postTaskEndEvent(TaskResultLost, "executor-3"))
+    assert(tracker.getExecutorBlacklist === Set("executor-1", "executor-2", "executor-3"))
+    assert(executorIdToTaskFailures(tracker)("executor-3").numFailures === 7)
+    assert(executorIdToTaskFailures(tracker)("executor-3").isBlackListed === true)
+
+    // Post TaskEnd event to executor-4 to verify whether executor-4 could be added into blacklist
+    (0 until 10).foreach(_ => postTaskEndEvent(TaskResultLost, "executor-4"))
+    // Event executor-4's failure task number is above than blacklist threshold,
+    // but the blacklisted executor number is reaching to maximum fraction,
+    // so executor-4 still cannot be added into blacklist.
```

Comment:
Yes, that's a design trade-off between fast fail and fail straggling. I'm not clearly sure which way is a better way.